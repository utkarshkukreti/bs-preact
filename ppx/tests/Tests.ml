module Preact = struct
  type undefined

  let undefined : undefined = Obj.magic ()

  let useState _ _ = Obj.magic ()

  let useReducer _ _ _ = Obj.magic ()

  let component _ _ = Obj.magic ()
end

let _foo =
 fun [@preact.component "Foo"] _ ->
  let[@hook] _count, _setCount = Preact.useState () in
  ()

let useThing =
 fun [@preact.hook] _ ->
  let[@hook] count, _setCount = Preact.useState () in
  let () = Js.log "in useThing!" in
  let[@hook] value, _dispatch = Preact.useReducer () () in
  count, value

let _foo =
 fun [@preact.component "Foo"] _ ->
  let[@hook] _count, _setCount = Preact.useState () in
  let[@hook] _value, _dispatch = Preact.useReducer () () in
  let[@hook] _, _ = useThing () in
  ()

(* These should produce errors. *)

(* let _foo _ =
  let[@hook] _ = () in
  () *)

(* let _foo =
 fun [@preact.component "Foo"] _ ->
  let[@hook] _count, _setCount =
    if true then Preact.useState () else Preact.useState ()
  in
  () *)

(* let _foo =
 fun [@preact.component "Foo"] _ ->
  if true
  then (
    let[@hook] _count, _setCount = Preact.useState () in
    ())
  else () *)

(* let _foo =
 fun [@preact.component "Foo"] _ ->
  let[@hook] () = () in
  () *)

(* let _foo =
 fun [@preact.component "Foo"] _ ->
  let[@hook] _count, _setCount = Preact.useState 0 in
  let[@hook] () = () in
  () *)

(* let _foo =
 fun [@preact.component "Foo"] _ ->
  let _ =
   fun () ->
    let[@hook] () = () in
    ()
  in
  () *)
