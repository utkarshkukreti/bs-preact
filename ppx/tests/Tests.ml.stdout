module Preact =
  struct
    type undefined
    let undefined: undefined = Obj.magic ()
    let useState _ _ = Obj.magic ()
    let useReducer _ _ _ = Obj.magic ()
    let component _ _ = Obj.magic ()
  end
let _foo =
  Preact.component "Foo"
    (fun _  ->
       let (_count,_setCount) = Preact.useState () Preact.undefined in ())
let useThing _ (_ : Preact.undefined) =
  let (count,_setCount) = Preact.useState () Preact.undefined in
  let () = Js.log "in useThing!" in
  let (value,_dispatch) = Preact.useReducer () () Preact.undefined in
  (count, value)
let _foo =
  Preact.component "Foo"
    (fun _  ->
       let (_count,_setCount) = Preact.useState () Preact.undefined in
       let (_value,_dispatch) = Preact.useReducer () () Preact.undefined in
       let (_,_) = useThing () Preact.undefined in ())
